Local API: Implement SABR for VODs

PikachuEXE opened the pull request:
<!-- Thanks for sending a pull request! Make sure to follow the contributing guidelines. -->
<!-- Important note, we may remove your pull request if you do not use this provided PR template correctly. -->
<!-- Do not create PR's with AI! (PRs created mainly with AI will be closed. They waste our team's time. We ban repeat offenders.) -->

## Pull Request Type
<!-- Please select what type of pull request this is: [x] -->
- [ ] Bugfix
- [x] Feature Implementation
- [ ] Documentation
- [ ] Other

## Related issue
<!-- Please link the issue your pull request is referring to. -->
<!-- If this pull request fully resolves the relevant issue, put "closes" before the issue number. -->
<!-- Example: "closes #123456". -->
https://github.com/FreeTubeApp/FreeTube/pull/7145
https://github.com/FreeTubeApp/FreeTube/pull/7887
Closes #6683
Closes #7119 
Closes #7920

## Description
<!-- Please write a clear and concise description of what the pull request does. -->
Based on https://github.com/FreeTubeApp/FreeTube/pull/7145 (Please read that PR first
Using https://github.com/LuanRT/googlevideo 4.0.4

- Implementation updated (different from #7145) to have playback session scoped stuff like SABR context & playbackCookie (SABR details)
- Player streaming timeout doubled (30s > 60s) to handle sometimes ~40MB response instead of ~10MB (at 2k quality, e.g. https://youtu.be/NmdXl7vXZ7g)
- New setting for enabling SABR in experimental section

Non-SABR
- Also updated the implementation for https://github.com/FreeTubeApp/FreeTube/pull/7887 to show pre-roll ads time in toast
- Toast now updated to be possible to show updated messages (hard coded to refresh per 1s currently)


## Screenshots <!-- If appropriate -->
<!-- Please add before and after screenshots if there is a visible change. -->
Lazy maybe add later?

## Testing
<!-- How can reviewers verify that the PR produces correct results? -->
<!-- Please provide instructions so that others can ensure that your pull request would produce correct results. For examples see, https://github.com/FreeTubeApp/FreeTube/pull/5743, https://github.com/FreeTubeApp/FreeTube/pull/7349, https://github.com/FreeTubeApp/FreeTube/pull/5125, https://github.com/FreeTubeApp/FreeTube/pull/7338 -->

### A. SABR off
- Test pre-roll ads time toast
- Test various video playback (should be the same)

### B. SABR on
- Test various video playback (should be the same)

Find custom builds at https://github.com/PikachuEXE/FreeTube/actions/workflows/build.yml?query=branch%3Asabr-pr
Find cutting edge custom builds (I only use this branch for testing fixes myself, but can be outdated once I finished testing): https://github.com/PikachuEXE/FreeTube/actions/workflows/build.yml?query=branch%3Acustom-builds%2Fsabr

## Desktop
<!-- Please complete the following information-->
- **OS:**
- **OS Version:**
- **FreeTube version:**

## Additional context
<!-- Add any other context about the pull request here. -->
Remaining tasks
- [x] Add/Update labels/tooltips/toasts & make them translatable (after text confirmed)
- [x] Handle infinite player reload - Allow max 3 player reloads for the same video before fallback to older DASH, if player loaded the reload doesn't count
- [x] Handle player reload during playback - Reload while keeping timestamp
- [x] Handle SABR playback delay in the middle - keep toast for now
- [x] Handle infinite SABR playback delay - reload player after (per req) backoff requested 5 times / cumulative delay going to be larger than timeout for request
- [x] Handle infinite next request policy received - reload player after (per req) next request policy received 100 times
- [x] Enable SABR by default
- [ ] Handle more edge cases if found

### Post PR
- Remove download feature (broken anyway)

PikachuEXE committed:
Local API: Implement SABR for VODs

efb4f5ff-1298-471a-8973-3d47447115dc commented:
Didnt test yet but im very sure this will close the SABR related issues

#7119
#6683

absidue commented:
I really hope it being an experimental setting that is disabled by default is just because you have some things left in the remaining tasks section...

PikachuEXE committed:
* Allow max 3 player reloads for the same video before fallback to older DASH

PikachuEXE committed:
! Fix data type for `playerTimeMs` (googlevideo 4.0.4)

PikachuEXE committed:
* Update player reload handling to restart player at the current timestamp

PikachuEXE committed:
* Do not count reload if player is loaded after that

efb4f5ff-1298-471a-8973-3d47447115dc reviewed (commented):
Both with and without SABR i get this error but that is on a server that is running through tor. If i connect to a normal server it does work

https://github.com/user-attachments/assets/a2e054d0-1b65-4b63-a8c2-9f8f5cda2b18

PikachuEXE committed:
! Fix access from nullable object

PikachuEXE committed:
* Remove code for debug only

PikachuEXE committed:
! Workaround infinite backoff (mostly at the middle of playback

PikachuEXE committed:
! Workaround infinite retry due to nextRequestPolicy (mostly at the middle of playback

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development: (24 commits)
  Add proxy login support (#7996)
  Migrate the History view to the composition API (#8064)
  Fix player button hiding (#8065)
  Translated using Weblate (Norwegian Bokmål)
  Translated using Weblate (Indonesian)
  Fix playback rate reset when video ends (#7718)
  Skip Electron binary download in CI as we do not run it (#8063)
  Feat: Display player overlay controls on keyboard interaction (#7757)
  Translated using Weblate (Hungarian)
  Translated using Weblate (Czech)
  Translated using Weblate (Estonian)
  Translated using Weblate (French)
  Bump electron from 37.4.0 to 38.1.0 (#8025)
  Translated using Weblate (Dutch)
  Translated using Weblate (Portuguese (Brazil))
  Translated using Weblate (German)
  Translated using Weblate (Breton)
  Translated using Weblate (Italian)
  Translated using Weblate (Chinese (Simplified Han script))
  Support exporting single playlists as a list of video URLs (#8054)
  ...

# Conflicts:
#	src/renderer/components/ft-shaka-video-player/ft-shaka-video-player.js

PikachuEXE committed:
! Fix no audio track selection when SABR used

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development:
  Bump the eslint group across 1 directory with 4 updates (#8073)
  Bump electron from 38.1.0 to 38.1.2 (#8071)
  Bump sass from 1.92.1 to 1.93.0 (#8067)
  Bump lefthook from 1.13.0 to 1.13.1 (#8072)
  Bump swiper from 12.0.1 to 12.0.2 (#8070)
  Translated using Weblate (Italian)
  Translated using Weblate (Czech)
  Translated using Weblate (Hungarian)
  Translated using Weblate (German)
  Translated using Weblate (Estonian)
  Translated using Weblate (Breton)
  Translated using Weblate (German)
  Translated using Weblate (Chinese (Traditional Han script))
  Improve playlist import performance (#8062)
  Translated using Weblate (Norwegian Bokmål)
  Translated using Weblate (Portuguese (Brazil))
  Translated using Weblate (Chinese (Simplified Han script))
  Translated using Weblate (Japanese)
  Handle lockup view videos without a view count (#8066)

PikachuEXE committed:
! Fix no audio track selection when SABR used with less requests

PikachuEXE committed:
* Use Math.round instead of Math.trunc for start time / duration like googlevideo repo

PikachuEXE committed:
* Add requestNumber to SABR URL

PikachuEXE committed:
! Fix SABR redirect handling

PikachuEXE committed:
! Fix reload sometimes not working

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development:
  Translated using Weblate (Turkish)
  Translated using Weblate (English (United Kingdom))
  Bump eslint-plugin-jsdoc from 60.2.0 to 60.5.0 in the eslint group (#8091)
  Bump sass from 1.93.1 to 1.93.2 (#8090)
  Bump lefthook from 1.13.3 to 1.13.4 (#8088)
  Bump electron from 38.1.2 to 38.2.0 (#8089)
  Migrate the SearchPage component to the composition API (#8086)
  Hide webpack-dev-server error overlay for application errors (#8087)
  Translated using Weblate (Icelandic)
  Translated using Weblate (Hungarian)
  Translated using Weblate (Swedish)
  ! Fix unable to add video to any playlist (#8082)
  Translated using Weblate (Polish)
  Translated using Weblate (Basque)
  Fix weblate parsing error (#8079)
  Bump sass from 1.93.0 to 1.93.1 (#8078)
  Bump lefthook from 1.13.1 to 1.13.3 (#8077)
  Bump eslint-plugin-jsdoc from 60.1.1 to 60.2.0 in the eslint group (#8076)
  Bump shaka-player from 4.16.1 to 4.16.2 (#8069)
  Added translation using Weblate (Marathi)

PikachuEXE committed:
* Reload with less cumulative backoff requested (5 > 3)

PikachuEXE committed:
* Implement poToken fallback (visitorData bind > videoID bind

PikachuEXE committed:
! Fix progress saving run for video reloading in the middle

PikachuEXE committed:
* Implement proper poToken selection

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development:
  Clean up upstreamed big play button opacity (#8052)
  Bump the eslint group with 5 updates (#8095)
  Bump stylelint from 16.24.0 to 16.25.0 in the stylelint group (#8096)
  Bump lefthook from 1.13.4 to 1.13.6 (#8099)
  Bump electron from 38.2.0 to 38.2.1 (#8100)
  Bump shaka-player from 4.16.2 to 4.16.3 (#8098)
  Bump webpack from 5.101.3 to 5.102.0 (#8097)
  Translated using Weblate (Hungarian)
  Translated using Weblate (Welsh)
  Change Player icons (#8057)
  Serialize search and playlist continuations in the store (#8092)
  Translated using Weblate (Italian)

PikachuEXE commented:
Latest build, just merged dev and with proper commits (i.e. only debug/workaround changes in a separate branch for this build, others all in this PR
Will make the new setting enabled by default next week if no major issues with playback (I expect you to test now if not started yet~

PikachuEXE commented:
Before I forgot: It seems what player reload will be requested (when SABR enabled) once poToken expires
I got this once I forgot to close a window and try to resume (and it starts buffering
Not sure if works every time though

PikachuEXE committed:
- Remove accidentially introduced debug code

PikachuEXE committed:
* skip the call to concatenateChunks if it isn't needed

PikachuEXE committed:
* Make changes for VS Code

PikachuEXE committed:
* Update new toast messaages to be translatable

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development:
  Update next video countdown toast to be refreshed per second instead of using workaround (#8103)
  Remove unused supportsYtdlProtocol from the external player map file (#8107)
  Translated using Weblate (Dutch)
  Translated using Weblate (Spanish)
  Translated using Weblate (Spanish)
  Translated using Weblate (Spanish)
  Translated using Weblate (Hebrew)
  Hide Proxy password (#8102)

# Conflicts:
#	src/renderer/components/FtToast/FtToast.vue

PikachuEXE committed:
no message

efb4f5ff-1298-471a-8973-3d47447115dc committed:
* Enable SABR by default

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development:
  Translated using Weblate (German)
  Fix importing and exporting data after the Electron 38.2.0 update (#8106)
  Translated using Weblate (Breton)
  Translated using Weblate (Polish)
  Translated using Weblate (Polish)
  Translated using Weblate (Icelandic)
  Translated using Weblate (Italian)
  Translated using Weblate (Chinese (Traditional Han script))
  Bump electron from 38.2.1 to 38.2.2 (#8112)
  Bump eslint-plugin-jsdoc from 60.8.2 to 61.1.1 in the eslint group (#8110)

PikachuEXE committed:
! Fix null endSegmentIndex handling

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development:
  Bump package version from 0.23.11 to 0.23.12 (#8138)
  Add error handling to the deciphering code (#8139)
  Fix video playback by using video ID bound poToken (#8137)
  Hide subscriptions wrapper on mobile and center bottom SideNav items (#8133)
  ! Fix sigFrameScript on dev in windows (#8134)
  Translated using Weblate (Estonian)
  Bump github/codeql-action from 3 to 4 (#8124)
  Translated using Weblate (Welsh)
  Bump mikefarah/yq from 4.47.2 to 4.48.1 (#8123)
  Bump stefanzweifel/git-auto-commit-action from 6 to 7 (#8125)
  Bump package version from 0.23.10 to 0.23.11 (#8126)

# Conflicts:
#	src/renderer/helpers/api/local.js

PikachuEXE committed:
* Load less data when requesting video data

PikachuEXE committed:
! Workaround player stuck at the end on windows

Due to duration somehow being slightly longer than what's available

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development:
  Add missing IS_ELECTRON and SUPPORTS_LOCAL_API checks (#8147)
  Simplify the NATIVE_THEME_UPDATE event handler (#8140)

PikachuEXE commented:
Maybe last build - https://github.com/PikachuEXE/FreeTube/actions/runs/18607715883
Fixed issues in https://github.com/PikachuEXE/FreeTube/issues/83 (maybe OS specific no idea)
Added a tweak to load less data for video/audio requests (no idea how to test if it really loads less data) taken from https://github.com/LuanRT/googlevideo/blob/googlevideo-v4.0.4/src/core/SabrStreamingAdapter.ts#L426

ArthurKun21 commented:
Hello, thanks for the work. Just wanna say for future reference.

The issue in https://github.com/PikachuEXE/FreeTube/issues/83 like the video getting stuck in the link https://youtu.be/8sJCPuweieA was present in both windows and mac machine. But the issue was already fixed with the succeeding commits

pnaf commented:
I just wanted to add that the stuck player issue in that link was also happening on Linux (I was also encountering the issue at the end of https://youtu.be/vBynw9Isr28, for example). So it seems less likely that this is an OS-specific issue.

The new build with the workaround did fix that issue for me. Thanks!

pnaf commented on src/renderer/helpers/player/SabrManifestParser.js:
```diff
@@ -0,0 +1,718 @@
+    presentationTimeline.setSegmentAvailabilityDuration(Infinity)
+    presentationTimeline.lockStartTime()
+    // Avoid player stuck at the end
+    // Reproducible on windows with some videos
```
```suggestion
    // Reproducible on multiple OSes with some videos
```

PikachuEXE committed:
~ Update comment about workaround

PikachuEXE committed:
! Fix player stuck at the end of video with shortest duration

due to use of duration from first format with format of shorter duration selected for playback

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development:
  Bump the eslint group with 4 updates (#8158)
  Translated using Weblate (German)
  Bump shaka-player from 4.16.4 to 4.16.6 (#8167)
  Bump marked from 16.4.0 to 16.4.1 (#8159)
  Bump electron from 38.2.2 to 38.3.0 (#8161)
  Bump actions/setup-node from 5 to 6 (#8162)
  Bump youtubei.js from 16.0.0 to 16.0.1 (#8163)
  Translated using Weblate (Persian)
  Translated using Weblate (Czech)
  Translated using Weblate (Bulgarian)

# Conflicts:
#	package.json

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development:
  Translated using Weblate (Russian)
  Bump the eslint group with 2 updates (#8184)
  Bump the babel group with 2 updates (#8185)
  Bump sass-loader from 16.0.5 to 16.0.6 (#8186)
  Bump electron from 38.3.0 to 38.4.0 (#8187)
  Bump swiper from 12.0.2 to 12.0.3 (#8189)
  Bump actions/upload-artifact from 4 to 5 (#8188)
  Bump lefthook from 2.0.0 to 2.0.1 (#8190)
  Handle lockup views with no published text or views (#8170)
  Translated using Weblate (Russian)
  Translated using Weblate (Russian)
  Remove Hide Channel from the kebab menu on the Subscriptions page (#8175)
  Translated using Weblate (French)
  Translated using Weblate (Latvian)
  Translated using Weblate (Latvian)
  Translated using Weblate (Portuguese (Portugal))
  Bump lefthook from 1.13.6 to 2.0.0 (#8166)

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development: (21 commits)
  Replace vue-portal dependency with Vue's built-in <Teleport> component (#8207)
  Update to Vue 3 (#8094)
  Translated using Weblate (Basque)
  Translated using Weblate (Hungarian)
  Translated using Weblate (Icelandic)
  Translated using Weblate (Turkish)
  Properly fix CORS in PO token web views (#8203)
  Translated using Weblate (Italian)
  Translated using Weblate (Norwegian Bokmål)
  Translated using Weblate (French)
  Configure nedb to always clean up corrupted data (#8202)
  Translated using Weblate (Portuguese (Brazil))
  Add video paused check to frame shortcuts (#8200)
  Fix toast with timeout 0 will be displayed for default 3s (#8168)
  Translated using Weblate (Chinese (Simplified Han script))
  Change containing text input limit (#8174)
  Translated using Weblate (Dutch)
  Add support for AppImageUpdate (#8153)
  Added New Window option to Tray (#7995)
  Translated using Weblate (Japanese)
  ...

PikachuEXE commented:
@kommunarr Sorry I can't reproduce SABR loop (reload loop?). It just starts playing right after backoff time
I have no idea if it's random or what (it's rare and doesn't happen again with the same video

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development:
  Translated using Weblate (German)
  Translated using Weblate (German)
  Bump shaka-player from 4.16.6 to 4.16.8 (#8223)
  Translated using Weblate (Croatian)
  Bump globals from 16.4.0 to 16.5.0 (#8216)
  Bump eslint-plugin-jsdoc from 61.1.11 to 61.1.12 in the eslint group (#8222)
  Translated using Weblate (Czech)
  Translated using Weblate (Estonian)
  Bump lefthook from 2.0.1 to 2.0.2 (#8213)
  Bump sass from 1.93.2 to 1.93.3 (#8214)
  Bump the eslint group with 3 updates (#8211)
  Remove broken restart window keyboard shortcut (#7281)
  Translated using Weblate (Vietnamese)
  Translated using Weblate (Hebrew)
  Translated using Weblate (Chinese (Traditional Han script))
  Migrate FtInput to the composition API (#8208)

PikachuEXE committed:
! Avoid using `player` when it's null (on reload)

StevenNL2000 commented:
@kommunarr Are you sure the reloading is infinite and not just long but finite? I've been using this branch and I've noticed that it sometimes has to reload 3 times for me before the video starts playing. My educated guess is that this corresponds to YouTube trying to show 3 preroll ads.

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development:
  Change clear filter button from text to icon (#8205)
  Feat clickable playlist progress bar (#7782)
  Add Hide channels based on text (#7948)
  Clean up defunct shaka-player types patching (#8224)

SilentParrot commented:
I noticed that this [video](https://youtu.be/1XsS-dv4b1s) reloads a few times before ad preroll, with the following errors in console
```
renderer.js:2 [YOUTUBEJS][Parser]: ZU: Type mismatch, got ListItemView expected MenuServiceItem | MenuServiceItemDownload.
    at qO (renderer.js:2:1755273)
    at XO (renderer.js:2:1763857)
    at new QC (renderer.js:2:1549742)
    at XO (renderer.js:2:1764068)
    at ZO (renderer.js:2:1764352)
    at new tI (renderer.js:2:1550477)
    at XO (renderer.js:2:1764068)
    at new jN (renderer.js:2:1729160)
    at XO (renderer.js:2:1764068)
    at ZO (renderer.js:2:1764352)
renderer.js:2 [YOUTUBEJS][Parser]: ZU: Type mismatch, got ExpandableMetadata expected VideoDescriptionHeader | ExpandableVideoDescriptionBody | VideoDescriptionMusicSection | VideoDescriptionInfocardsSection | VideoDescriptionCourseSection | VideoDescriptionTranscriptSection | VideoDescriptionTranscriptSection | HorizontalCardList | ReelShelf | VideoAttributesSectionView | HowThisWasMadeSectionView.
    at qO (renderer.js:2:1755273)
    at XO (renderer.js:2:1763857)
    at ZO (renderer.js:2:1764352)
    at new tE (renderer.js:2:1570558)
    at XO (renderer.js:2:1764068)
    at new nE (renderer.js:2:1570795)
    at XO (renderer.js:2:1764068)
    at ZO (renderer.js:2:1764352)
    at YO (renderer.js:2:1762710)
    at new cU (renderer.js:2:1769382)```

PikachuEXE commented:
I can see some reload loop issue for random videos but after a few hours it's gone
Wondering if it's some YT protection/rate limiting/unknown logic
I can reproduce with your video but no idea how to fix yet

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development: (41 commits)
  Translated using Weblate (Turkish)
  ! Fix top nav cannot search in new window (#8254)
  Bump vue from 3.5.22 to 3.5.24 (#8257)
  Translated using Weblate (Czech)
  Translated using Weblate (Italian)
  Translated using Weblate (Polish)
  Translated using Weblate (Hungarian)
  Translated using Weblate (Norwegian Bokmål)
  Bump lefthook from 2.0.2 to 2.0.3 (#8258)
  Bump the eslint group with 2 updates (#8256)
  Bump marked from 16.4.1 to 17.0.0 (#8260)
  Translated using Weblate (Breton)
  Translated using Weblate (Estonian)
  Translated using Weblate (French)
  Translated using Weblate (Chinese (Simplified Han script))
  Add support for importing and exporting search history (#8237)
  Invidious API: Fix published dates on recommended videos (#8251)
  Translated using Weblate (Chinese (Simplified Han script))
  ! Fix playlist progress bar preview broken in IV API (#8245)
  Wrap interactive FontAwesomeIcons in actual buttons (#8206)
  ...

# Conflicts:
#	package.json

PikachuEXE committed:
~ Fix type definition

PikachuEXE committed:
! Workaround player reload issue

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development:
  Fix playlist video view counts on the local API (#8290)
  change hardcoded color app text logoColor (#8288)
  Translated using Weblate (Russian)
  Translated using Weblate (German)
  wrap long playlist strings (#8275)
  Translated using Weblate (Bulgarian)
  Bump js-yaml from 4.1.0 to 4.1.1 (#8284)
  bump workflows node-version to 24 (#8279)
  Translated using Weblate (Basque)
  Translated using Weblate (Vietnamese)
  Translated using Weblate (Hungarian)
  Translated using Weblate (Hungarian)
  Remove upperCase on various places (#8267)
  Translated using Weblate (French)
  Fix subscription count display on watch page with IV API (#8255)
  Specify window position in constructor for better wayland compatibility (#8238)
  Change profile icon to include lowercase letters (#8253)
  Translated using Weblate (German)
  Translated using Weblate (Chinese (Traditional Han script))
  Translated using Weblate (Portuguese (Brazil))

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development:
  Translated using Weblate (Chinese (Simplified Han script))
  Workaround ready-to-show not firing consistently on wayland (#8294)
  Bump electron from 38.4.0 to 39.2.1 (#8303)
  Migrate the WatchVideoInfo component to the composition API (#8295)
  Translated using Weblate (Hungarian)
  Translated using Weblate (Norwegian Bokmål)
  Translated using Weblate (French)
  Bump eslint-plugin-jsdoc from 61.1.12 to 61.2.1 in the eslint group (#8298)
  Replace broken trending with a local API only equivalent (#8289)
  Fix settings layout issues after the Vue 3 migration (#8292)
  Bump lefthook from 2.0.3 to 2.0.4 (#8299)
  Bump shaka-player from 4.16.8 to 4.16.9 (#8301)
  Bump electron-builder from 26.1.0 to 26.2.0 (#8302)

# Conflicts:
#	src/renderer/helpers/api/local.js

PikachuEXE committed:
* Make new setting related text translatable & update text

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development: (72 commits)
  Bump shaka-player from 4.16.9 to 4.16.10 (#8345)
  Bump the stylelint group with 2 updates (#8334)
  Bump the eslint group with 2 updates (#8333)
  Bump vue from 3.5.24 to 3.5.25 (#8336)
  Bump electron-builder from 26.2.0 to 26.3.0 (#8341)
  Bump marked from 17.0.0 to 17.0.1 (#8339)
  Bump webpack from 5.102.1 to 5.103.0 (#8340)
  Translated using Weblate (English (United Kingdom))
  Bump html-webpack-plugin from 5.6.4 to 5.6.5 (#8342)
  Bump mikefarah/yq from 4.48.1 to 4.49.1 (#8343)
  Bump actions/checkout from 5 to 6 (#8344)
  Bump electron from 39.2.1 to 39.2.3 (#8337)
  Bump sass from 1.94.0 to 1.94.2 (#8335)
  Translated using Weblate (Dutch)
  Translated using Weblate (Chinese (Traditional Han script))
  Fix dropdownShown error on the watch page for live streams (#8331)
  Fix view count with RSS feeds (#8328)
  Add support for exporting watch history in YouTube's JSON format (#8323)
  Cleanup leftover import toast (#8325)
  Translated using Weblate (Portuguese (Brazil))
  ...

# Conflicts:
#	package.json

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development: (50 commits)
  Translated using Weblate (Slovak)
  Translated using Weblate (Slovak)
  Translated using Weblate (Slovak)
  add skip to previous/next video buttons in media player (#8285)
  Translated using Weblate (Slovak)
  Translated using Weblate (Slovak)
  Translated using Weblate (German)
  Bump express from 4.21.2 to 4.22.0 (#8374)
  Bump electron-builder from 26.3.0 to 26.3.2 (#8370)
  Translated using Weblate (Hebrew)
  Bump the eslint group with 2 updates (#8367)
  Translated using Weblate (Portuguese)
  Bump stylelint from 16.26.0 to 16.26.1 in the stylelint group (#8368)
  Bump vue-i18n from 11.1.12 to 11.2.2 (#8369)
  Bump electron from 39.2.3 to 39.2.4 (#8371)
  Bump mikefarah/yq from 4.49.1 to 4.49.2 (#8372)
  Translated using Weblate (Slovak)
  Translated using Weblate (Slovak)
  Translated using Weblate (Slovak)
  Translated using Weblate (Slovak)
  ...

efb4f5ff-1298-471a-8973-3d47447115dc reviewed (commented):
Created a test playlist with a mix of every type of video in it. I've received allot of reloads and backoff time varies from 4-5 sec to 14-15sec.

Just to be clear im not against the backoff time but when a video reloads and let me wait again and again it just doesnt feel right

 Im not sure if i feel comfortable with this annoyance level :/. I understand the choice to fallback to non-SABR a bit more now but like i said in previous conversations we shouldnt include non SABR options at all

https://github.com/user-attachments/assets/b377603d-7042-4c32-97c4-4a4a2a1d83b2

https://github.com/user-attachments/assets/2194ae69-e44a-46c6-b033-2e802c669d2b

https://github.com/user-attachments/assets/479ecb44-77b8-467e-a9a7-ebead3fa87e8

https://github.com/user-attachments/assets/f7f1400e-872c-4aa7-9faf-dd3a6bf98751

https://github.com/user-attachments/assets/29b8c4eb-1c5d-4647-9dbe-70c18238c18e

https://github.com/user-attachments/assets/1ad870bb-5ef2-426e-9e36-35e21a62deea

https://github.com/user-attachments/assets/a2d41b4f-d8af-4c6a-ab02-62a435d407c5

https://github.com/user-attachments/assets/38d62654-74b3-4cf7-bfda-e2805b987f33

https://github.com/user-attachments/assets/7164ef44-2167-46e7-ba72-99a454735bbf

https://github.com/user-attachments/assets/d966646d-422c-4bd0-a136-f090c07d9bed

absidue commented:
Last time I checked webpack only validates imports against the `exports` field in a package's `package.json` file if you use a package import e.g. `import x from 'youtubei.js'` but if you instead specify the full relative path into the `node_modules` folder to that file, you can import any file you want. So you could import the `FormatXTags` protobuf decoder and detect the voice boost formats in FreeTube without waiting for a YouTube.js update.

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development: (62 commits)
  Support exporting single playlists in YouTube's CSV format (#8430)
  Move external player handling to the main process (#8416)
  Revert "[Bug fix] Set max-block-size to prevent thin videos from passing the …" (#8418)
  Migrate the App component to the composition API (#8417)
  Bump electron-builder from 26.3.5 to 26.3.6 (#8422)
  Bump css-minimizer-webpack-plugin from 7.0.3 to 7.0.4 (#8428)
  Bump the eslint group with 3 updates (#8420)
  Bump sass from 1.94.2 to 1.96.0 (#8429)
  Bump vue-router from 4.6.3 to 4.6.4 (#8425)
  Bump peter-evans/create-pull-request from 7 to 8 (#8423)
  Bump mikefarah/yq from 4.49.2 to 4.50.1 (#8427)
  Bump lefthook from 2.0.8 to 2.0.11 (#8426)
  Bump electron from 39.2.6 to 39.2.7 (#8421)
  Bump actions/upload-artifact from 5 to 6 (#8424)
  Translated using Weblate (Russian)
  Translated using Weblate (Bulgarian)
  Translated using Weblate (German)
  Translated using Weblate (Indonesian)
  Translated using Weblate (Portuguese (Brazil))
  Translated using Weblate (Japanese)
  ...

PikachuEXE committed:
Merge branch 'development' into sabr-pr

* development: (26 commits)
  Use bash even in workflows that run on Windows (#8493)
  Translated using Weblate (Malay)
  fix refresh (#8475)
  Harden GitHub Actions workflows (#8476)
  Update copyright year to 2026 (#8487)
  Added translation using Weblate (Malay)
  Bump globals from 16.5.0 to 17.0.0 (#8484)
  Bump vue-i18n from 11.2.7 to 11.2.8 (#8483)
  Translated using Weblate (Welsh)
  Bump qs from 6.14.0 to 6.14.1 (#8471)
  Translated using Weblate (Estonian)
  Bump lefthook from 2.0.12 to 2.0.13 (#8461)
  Translated using Weblate (Dutch)
  Fix DeArrow thumbnails for shorts getting stretched (#8434)
  Clean up FtSlider CSS (#8438)
  Migrate the GeneralSettings component to the composition API (#8439)
  Translated using Weblate (English (United Kingdom))
  Bump vue-i18n from 11.2.2 to 11.2.7 (#8451)
  Bump electron-builder from 26.3.6 to 26.4.0 (#8446)
  Bump lefthook from 2.0.11 to 2.0.12 (#8449)
  ...

absidue committed:
Merge branch 'development' into sabr-pr

absidue committed:
Add support for Voice Boost formats

MulverineX commented:
I gave this a shot and its working for me, haven't tested it that extensively though. I'm curious, does the audio-only stream get SABR to only send audio? Or is the video content in the downloaded chunks simply ignored when audio-only mode is enabled?

efb4f5ff-1298-471a-8973-3d47447115dc commented:
working perfectly like reported earlier internally but when i set default formats to legacy in the settings it always plays dash

PikachuEXE commented:
I got this and I have no idea how to fix yet
Update 1: No such issue on dev
<img width="596" height="960" alt="image" src="https://github.com/user-attachments/assets/f0eb7ac0-ac9a-45c8-9d86-906b9df4166a" />

PikachuEXE committed:
! Fix legacy format handling

absidue committed:
Fix voice boost being selected as the default audio track

efb4f5ff-1298-471a-8973-3d47447115dc commented:
New build: https://github.com/PikachuEXE/FreeTube/actions/runs/21094847809

efb4f5ff-1298-471a-8973-3d47447115dc commented:
remaining thing todo is to get rid of the toggle in experimental settings and the fallback to the workarounds

ChunkyProgrammer reviewed (approved):
Great work everyone ❤️
